### PR TITLE
Make it build on Macos and fix how you template a container template

### PR DESCRIPTION
**Pull request by SokolAndrey**
Successfully built and tested from the source code following the instructions in the README on
```
uname -msr
Darwin 23.6.0 arm64
```
Also tested the build on a Ubuntu machine with a simple script which encodes and decodes random text.

### CODE REVIEW

File: src/spm_bpe_cache_merge.cc [hunk 1]
```diff
@@ -295,10 +296,10 @@ int main(int argc, char *argv[]) {
     LOG(INFO) << merged.sentences_size << " sentences, "
               << merged.required_chars.size() << " chars; allocated "
               << merged.allocated();
-    #ifdef TCMALLOC
+    #if defined(TCMALLOC)
     MallocExtension::instance()->ReleaseFreeMemory();
-    #endif
     malloc_stats();
```
Comment (vmarkovtsev) [inline src/spm_bpe_cache_merge.cc:301]: `malloc_stats()` doesn't depend on TCMALLOC and requires `malloc.h`: https://man7.org/linux/man-pages/man3/malloc_stats.3.html

We shouldn't replace `malloc.h` with `stdlib.h`. Instead, all the parts of the code that depend on `malloc.h` must be skipped on macOS:
```
#ifdef __GLIBC__
#include <malloc.h>
#endif

#ifdef __GLIBC__
malloc_stats();
#endif
```

File: src/bpe_model_trainer.cc [hunk 2]
```diff
@@ -14,7 +14,8 @@
 
 #include "bpe_model_trainer.h"
 
-#include <malloc.h>
+
+#include <stdlib.h>
```
Comment (vmarkovtsev) [inline src/bpe_model_trainer.cc:18]: Same here

File: src/trainer_interface.cc [hunk 3]
```diff
@@ -14,8 +14,9 @@
 
 #include "trainer_interface.h"
 
-#include <malloc.h>
-#ifdef TCMALLOC
+#include <stdlib.h>
```
Comment (vmarkovtsev) [inline src/trainer_interface.cc:17]: Same here

File: src/util.h [hunk 4]
```diff
@@ -289,7 +289,7 @@ namespace random {
 
 std::mt19937 *GetRandomGenerator();
 
-template <typename T, template<class> typename C>
+template <typename T, template <typename...> class C>
```
Comment (vmarkovtsev) [inline src/util.h:292]: We must check that this doesn't break gcc on Linux, in return.

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

File: src/util.h (hunk 4 above)
Comment (SokolAndrey) [reply src/util.h:292]: could you pls give me some direction how we should test it besides building on a linux machine?

Comment (vmarkovtsev) [reply src/util.h:292]: Just building is enough. Once it compiles, it should work!

**Review by vmarkovtsev — APPROVED** [review]
LGTM

**Pull request merged**